libr/core/cmd_debug.c

```diff
@@ -1217,103 +1184,191 @@ static bool print_double_linked_list_bin(RCore *core,  RHeap_MallocState *main_a
 	next = (ut64)(size_t)cnk->fd;
 	if (next != bin) {
 		eprintf ("Double linked list corrupted\n");
-		return false;
+		return -1;
 	}
 	r_core_read_at (core, next, (ut8 *)cnk, sizeof (RHeapChunk));
-
 	PRINTF_GA ("\n    0x%"PFMT64x, (ut64)(size_t) bin);
 	while ((ut64)(size_t)cnk->bk != bin) {
 		PRINTF_BA ("->bk = 0x%"PFMT64x, (ut64)(size_t)(cnk->bk));
-		next = (ut64)(size_t)cnk->bk;
+		next = (size_t)cnk->bk;
 		if (next < brk_start || next > (ut64)(size_t)main_arena->top) {
 			eprintf ("Double linked list corrupted\n");
-			return false;
+			return -1;
 		}
 		r_core_read_at (core, next, (ut8 *)cnk, sizeof (RHeapChunk));
 	}
 
 	PRINTF_GA ("->bk = 0x%"PFMT64x, (ut64)(size_t)(cnk->bk));
-	PRINT_GA ("\n  }\n");
+	return 1;
+}
+
+static int print_double_linked_list_bin_graph(RCore *core, ut64 bin, RHeap_MallocState *main_arena, ut64 brk_start) {
+	RAGraph *g = r_agraph_new (r_cons_canvas_new (1, 1));
+	g->can->color = r_config_get_i (core->config, "scr.color");
+
+	ut64 next = UT64_MAX;
+	char title[256] = {0}, chunk[256] = {0};
+	RANode *bin_node = NULL, *prev_node = NULL, *next_node = NULL;	
+	RHeapChunk *cnk = R_NEW0 (RHeapChunk);
+	if (!cnk) {
+		eprintf ("Warning: out of memory\n");
+		return 0;
+	}
+
+	r_core_read_at (core, bin, (ut8 *)cnk, sizeof (RHeapChunk));
+	snprintf (title, sizeof (title) - 1, "bin @ 0x%"PFMT64x"\n", (ut64)(size_t)bin);
+	snprintf (chunk, sizeof (chunk) - 1, "fd: 0x%"PFMT64x"\nbk: 0x%"PFMT64x"\n", (ut64)(size_t)cnk->fd, (ut64)(size_t)cnk->bk);
+	bin_node = r_agraph_add_node (g, title, chunk);
+	prev_node = bin_node;
+
+	while ((ut64)(size_t)cnk->bk != bin) {
+		next = (ut64)(size_t)cnk->bk;
+		if (next < brk_start || next > (ut64)(size_t)main_arena->top) {
+			PRINT_RA ("Double linked list corrupted\n");
+			return -1;
+		}	
+		r_core_read_at (core, next, (ut8 *)cnk, sizeof (RHeapChunk));
+		snprintf (title, sizeof (title) - 1, "Chunk @ 0x%"PFMT64x"\n", (ut64)(size_t)next);
+		snprintf (chunk, sizeof (chunk) - 1, "fd: 0x%"PFMT64x"\nbk: 0x%"PFMT64x"\n", (ut64)(size_t)cnk->fd, (ut64)(size_t)cnk->bk);
+		next_node = r_agraph_add_node (g, title, chunk);
+		r_agraph_add_edge (g, prev_node, next_node);
+		r_agraph_add_edge (g, next_node, prev_node);
+		prev_node = next_node;
+	}
+	
+	r_agraph_add_edge (g, prev_node, bin_node);
+	r_agraph_add_edge (g, bin_node, prev_node);
+	r_agraph_print (g);
+
 	free (cnk);
+	r_agraph_free (g);
 
-	return true;
+	return 1;
+}
+
+static int print_double_linked_list_bin(RCore *core,  RHeap_MallocState *main_arena, ut64 m_arena, ut64 offset, ut64 num_bin, int graph) {
+	int ret = 0;
+	RListIter *iter;
+	RDebugMap *map;
+	ut64 brk_start = UT64_MAX;
+	ut64 bin = (size_t)main_arena->bins[num_bin];
+	if (!bin) {
+		return ret;
+	}
+
+	if (!core || !core->dbg || !core->dbg->maps) {
+                return ret;
+	}
+
+	r_debug_map_sync (core->dbg);
+	r_list_foreach (core->dbg->maps, iter, map) {
+		if (strstr (map->name, "heap")) {
```

Comment:
Is maybe a better check for this? Maybe put this logic into a separate function?